--- a/perl/lib/Wallet/Object/Password.pm
+++ b/perl/lib/Wallet/Object/Password.pm
@@ -187,8 +187,7 @@ sub retrieve {
     if ($Wallet::Config::LDAP_SECRET) {
         my $undata = Wallet::Object::File->string_decrypt($data);
         if ($undata eq $data) {
-            my $endata = Wallet::Object::File->file_encrypt($data);
-            $self->_write_pw_file($path, $endata);
+            $self->_write_pw_file($path, $data);
             $self->log_action ($operation, $user, $host, $time);
         }
         $data = $undata;
